v2/routing_test.go

```diff
@@ -47,6 +51,87 @@ func makePkKeyValue(t *testing.T) (string, []byte) {
 	return routing.KeyForPublicKey(id), v
 }
 
+func TestDHT_FindPeer_happy_path(t *testing.T) {
+	ctx := kadtest.CtxShort(t)
+
+	top := NewTopology(t)
+	d1 := top.AddServer(nil)
+	d2 := top.AddServer(nil)
+	d3 := top.AddServer(nil)
+	d4 := top.AddServer(nil)
+	top.ConnectChain(ctx, d1, d2, d3, d4)
+
+	addrInfo, err := d1.FindPeer(ctx, d4.host.ID())
+	require.NoError(t, err)
+	assert.Equal(t, d4.host.ID(), addrInfo.ID)
+}
+
+func TestDHT_FindPeer_not_found(t *testing.T) {
+	ctx := kadtest.CtxShort(t)
+
+	top := NewTopology(t)
+	d1 := top.AddServer(nil)
+	d2 := top.AddServer(nil)
+	d3 := top.AddServer(nil)
+	d4 := top.AddServer(nil)
+	top.ConnectChain(ctx, d1, d2, d3)
+
+	_, err := d1.FindPeer(ctx, d4.host.ID())
+	assert.Error(t, err)
+}
+
+func TestDHT_FindPeer_already_connected(t *testing.T) {
+	ctx := kadtest.CtxShort(t)
+
+	top := NewTopology(t)
+	d1 := top.AddServer(nil)
+	d2 := top.AddServer(nil)
+	d3 := top.AddServer(nil)
+	d4 := top.AddServer(nil)
+	top.ConnectChain(ctx, d1, d2, d3)
+
+	err := d1.host.Connect(ctx, peer.AddrInfo{
+		ID:    d4.host.ID(),
+		Addrs: d4.host.Addrs(),
+	})
+	require.NoError(t, err)
+
+	_, err = d1.FindPeer(ctx, d4.host.ID())
+	assert.NoError(t, err)
+}
+
+func TestDHT_PutValue_happy_path(t *testing.T) {
+	// TIMING: this test is based on timeouts - so might become flaky!
+	ctx := kadtest.CtxShort(t)
+
+	top := NewTopology(t)
+	d1 := top.AddServer(nil)
+	d2 := top.AddServer(nil)
+
+	top.ConnectChain(ctx, d1, d2)
+
+	k, v := makePkKeyValue(t)
+
+	err := d1.PutValue(ctx, k, v)
+	require.NoError(t, err)
+
+	deadline, hasDeadline := ctx.Deadline()
+	if !hasDeadline {
+		deadline = time.Now().Add(5 * time.Second)
+	}
+
+	// putting data to a remote peer is an asynchronous operation. Even after
+	// PutValue returns, and although we have closed the stream on our end, an
+	// acknowledgement that the other peer has received the data is not
+	// guaranteed. The data will be flushed at this point, but the remote might
+	// not have handled it yet. Therefore, we use "EventuallyWithT" here.
+	assert.EventuallyWithT(t, func(t *assert.CollectT) {
```

Comment:
We control both peers in this test so can't we deterministically query the remote peer to see if the data has been handled?

Reply:
In the function body, I'm querying the remote peer to see if the data has been handled. Because of the reasons in the comment above I'm not sure how we can do this deterministically.

I could pass a custom Backend/Datastore to d2 and use a wait group to block until it's stored on the other peer 🤔